### PR TITLE
gh-114887 Reject only sockets of type SOCK_STREAM in create_datagram_endpoint(), improve exception message.

### PR DESCRIPTION
**Pull request by tjhowse**
Details in #114887.

* Issue: gh-114887

### CODE REVIEW

**Review by gvanrossum — COMMENTED** [review]
LGTM. Can you add a brief blurb? Then I'll merge it.

File: Misc/NEWS.d/next/Core and Builtins/2024-02-03-04-07-18.gh-issue-114887.uLSFmN.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Changed socket type validation in loop.create_datagram_endpoint() to accept all non-stream sockets. This fixes a regression in compatibility with raw sockets.
```
Comment (corona10) [inline Misc/NEWS.d/next/Core and Builtins/2024-02-03-04-07-18.gh-issue-114887.uLSFmN.rst]:
```suggestion
Changed socket type validation in :meth:`loop.create_datagram_endpoint` to accept all non-stream sockets.
This fixes a regression in compatibility with raw sockets.
```

Comment (corona10) [reply Misc/NEWS.d/next/Core and Builtins/2024-02-03-04-07-18.gh-issue-114887.uLSFmN.rst]: hmm it doesn't work.. :(

Comment (corona10) [reply Misc/NEWS.d/next/Core and Builtins/2024-02-03-04-07-18.gh-issue-114887.uLSFmN.rst]: Okay it works

```suggestion
Changed socket type validation in :meth:`~asyncio.loop.create_datagram_endpoint` to accept all non-stream sockets.
This fixes a regression in compatibility with raw sockets.
```

Comment (corona10) [reply Misc/NEWS.d/next/Core and Builtins/2024-02-03-04-07-18.gh-issue-114887.uLSFmN.rst]: 스크린샷 2024-02-03 오후 6 13 17

Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Core and Builtins/2024-02-03-04-07-18.gh-issue-114887.uLSFmN.rst]: I you with to keep "loop.":
```suggestion
Changed socket type validation in :meth:`loop.create_datagram_endpoint
<asyncio.loop.create_datagram_endpoint>` to accept all non-stream sockets.
This fixes a regression in compatibility with raw sockets.
```

**Comment by gvanrossum** [comment]
Is it worth bnackporting to 3.12 and 3.11?

**Comment by tjhowse** [comment]
> Is it worth backporting to 3.12 and 3.11?

I think so, as we're currently running 3.11.x in our application. PRs raised: #114977 and #114978

**Comment by tjhowse** [comment]
> > Is it worth backporting to 3.12 and 3.11?
> 
> I think so, as we're currently running 3.11.x in our application. PRs raised: #114977 and #114978

Hm, I seem to have botched the format of the branch names. New PRs raised: #114979 #114980

**Pull request merged**